rs-matter/src/data_model/core.rs

```diff
@@ -15,142 +15,759 @@
 
-// TODO: For now...
-static SUBS_ID: AtomicU32 = AtomicU32::new(1);
+use super::objects::*;
+use super::subscriptions::Subscriptions;
 
 /// The Maximum number of expanded writer request per transaction
 ///
 /// The write requests are first wildcard-expanded, and these many number of
 /// write requests per-transaction will be supported.
-pub const MAX_WRITE_ATTRS_IN_ONE_TRANS: usize = 7;
+const MAX_WRITE_ATTRS_IN_ONE_TRANS: usize = 7;
+
+pub type IMBuffer = heapless::Vec<u8, MAX_EXCHANGE_RX_BUF_SIZE>;
+
+struct SubscriptionBuffer<B> {
+    node_id: u64,
+    id: u32,
+    buffer: B,
+}
+
+/// An `ExchangeHandler` implementation capable of handling responder exchanges for the Interaction Model protocol.
+/// The implementation needs a `DataModelHandler` instance to interact with the underlying clusters of the data model.
+pub struct DataModel<'a, const N: usize, B, T>
+where
+    B: BufferAccess<IMBuffer>,
+{
+    handler: T,
+    subscriptions: &'a Subscriptions<N>,
+    subscriptions_buffers: RefCell<heapless::Vec<SubscriptionBuffer<B::Buffer<'a>>, N>>,
+    buffers: &'a B,
+}
+
+impl<'a, const N: usize, B, T> DataModel<'a, N, B, T>
+where
+    B: BufferAccess<IMBuffer>,
+    T: DataModelHandler,
+{
+    /// Create the handler.
+    #[inline(always)]
+    pub const fn new(buffers: &'a B, subscriptions: &'a Subscriptions<N>, handler: T) -> Self {
+        Self {
+            handler,
+            subscriptions,
+            subscriptions_buffers: RefCell::new(heapless::Vec::new()),
+            buffers,
+        }
+    }
+
+    /// Answer a responding exchange using the `DataModelHandler` instance wrapped by this exchange handler.
+    pub async fn handle(&self, exchange: &mut Exchange<'_>) -> Result<(), Error> {
+        let mut timeout_instant = None;
+
+        loop {
+            let mut repeat = false;
+
+            if exchange.rx().is_err() {
+                exchange.recv_fetch().await?;
+            }
+
+            let meta = exchange.rx()?.meta();
+            if meta.proto_id != PROTO_ID_INTERACTION_MODEL {
+                Err(ErrorCode::InvalidProto)?;
+            }
+
+            match meta.opcode::<OpCode>()? {
+                OpCode::ReadRequest => self.read(exchange).await?,
+                OpCode::WriteRequest => {
+                    repeat = self.write(exchange, timeout_instant.take()).await?;
+                }
+                OpCode::InvokeRequest => self.invoke(exchange, timeout_instant.take()).await?,
+                OpCode::SubscribeRequest => self.subscribe(exchange).await?,
+                OpCode::TimedRequest => {
+                    timeout_instant = Some(self.timed(exchange).await?);
+                    repeat = true;
+                }
+                opcode => {
+                    error!("Invalid opcode: {:?}", opcode);
+                    Err(ErrorCode::InvalidOpcode)?
+                }
+            }
 
-pub struct DataModel<T>(T);
+            if !repeat {
```

Comment:
nit: usage and setting of repeat spans a lot of code. have you considered having:

```rust
if meta.opcode::<OpCode>() != OpCode::TimedRequest {
    break;
}
```

and not have repeat at all? This makes variables have less of a span,

Reply:
But the loop should also repeat in the case where we have chunked writes. I.e. this code:
```rust
repeat = self.write(exchange, timeout_instant.take()).await?;
```
?

Reply:
I missed that bit of the code. never mind then.